b2sdk/transfer/emerge/executor.py

```diff
@@ -0,0 +1,480 @@
+    MAX_LARGE_FILE_SIZE = 10 * 1000 * 1000 * 1000 * 1000  # 10 TB
+
+    def __init__(
+        self,
+        session,
+        services,
+        bucket_id,
+        file_name,
+        content_type,
+        file_info,
+        progress_listener,
+        continue_large_file_id=None,
+        max_queue_size=None,
+    ):
+        super(LargeFileEmergeExecution, self).__init__(
+            session, services, bucket_id, file_name, content_type, file_info, progress_listener
+        )
+        self.continue_large_file_id = continue_large_file_id
+        self.max_queue_size = max_queue_size
+
+    def execute_plan(self, emerge_plan):
+        total_length = emerge_plan.get_total_length()
+
+        if total_length is not None and self.MAX_LARGE_FILE_SIZE < total_length:
+            raise MaxFileSizeExceeded(total_length, self.MAX_LARGE_FILE_SIZE)
+
+        plan_id = emerge_plan.get_plan_id()
+
+        file_info = dict(self.file_info) if self.file_info is not None else {}
+        if plan_id is not None:
+            file_info['plan_id'] = plan_id
+
+        self.progress_listener.set_total_bytes(total_length or 0)
+
+        emerge_parts_dict = None
+        if total_length is not None:
+            emerge_parts_dict = dict(emerge_plan.enumerate_emerge_parts())
+
+        unfinished_file, finished_parts = self._get_unfinished_file_and_parts(
+            self.bucket_id,
+            self.file_name,
+            file_info,
+            self.continue_large_file_id,
+            emerge_parts_dict=emerge_parts_dict
+        )
+
+        if unfinished_file is None:
+            if self.content_type is None:
+                content_type = self.DEFAULT_CONTENT_TYPE
+            else:
+                content_type = self.content_type
+            unfinished_file = self.services.large_file.start_large_file(
+                self.bucket_id, self.file_name, content_type, file_info
+            )
+        file_id = unfinished_file.file_id
+
+        semaphore = None
+        if self.max_queue_size is not None:
+            semaphore = threading.Semaphore(self.max_queue_size)
+
+        with self.progress_listener:
+            large_file_upload_state = LargeFileUploadState(self.progress_listener)
+
+            part_futures = []
+            for part_number, emerge_part in emerge_plan.enumerate_emerge_parts():
+                execution_step_factory = LargeFileEmergeExecutionStepFactory(
+                    self,
+                    emerge_part,
+                    part_number,
+                    file_id,
+                    large_file_upload_state,
+                    finished_parts=finished_parts,
+                )
+                execution_step = execution_step_factory.get_execution_step()
+                if semaphore is not None:
+                    semaphore.acquire()
+                future = execution_step.execute()
+                if semaphore is not None:
+                    future.add_done_callback(lambda f: semaphore.release())
```

Comment:
are you sure it is going to get called if the future is already done? Docs say it will be called in an event loop but are we sure it is going to be entered?

Can we abstract this away into something like b2sdk/bounded_queue_executor.py to separate execute_plan from this mechanic?